### PR TITLE
Allow to store and replay messages with context menu

### PR DESCRIPTION
**Pull request by tdraier**
fixes: [#1700](https://github.com/dust-tt/tasks/issues/1700)

## Description

Add an button on first message in a conversation to "save" as predefined conversation. Save these in local storage. Update context menu to recreate a new conversation with same message, including text and/or screenshot.

- Create new convo

  Screenshot 2024-11-25 at 17 34 01

- In conversation, click on "save " : 

  Screenshot 2024-11-25 at 17 34 24

- Right click  on any page : 

  Screenshot 2024-11-25 at 17 34 54

## To check

- Using a button in the first message was the easiest way to set-up these configs (and it;s actually quite easy to use), but we may want to do something else ?

- The item name in context menu is the entire user prompt, maybe need to put a limit to the length of the text

- We should change "Save to context menu" to something else, as these predefined configurations could  be displayed in home page of the extension (see "translate" and "summarize" buttons in figma design : 

  Screenshot 2024-11-25 at 17 48 28

- Do we want to have these buttons appear on the home page now ? If we have a lot, maybe a drop down would be more efficient ?

### CODE REVIEW

**Comment by pinotalexandre** [comment]
> Using a button in the first message was the easiest way to set-up these configs (and it's actually quite easy to use), but we may want to do something else ?

I'm ok with this pattern

>The item name in context menu is the entire user prompt, maybe need to put a limit to the length of the text

Yes sure, I'm proposing a 60 char text-length limit, wdyt?

>Do we want to have these buttons appear on the home page now ? If we have a lot, maybe a drop down would be more efficient ? 
I quite like the fact that we have theses two buttons as quick actions, I think it represent a lot of use case for the extension. Do you have other idea for this kind of button ? Otherwise, I don't think this is a priority for the browser extension. I'm ok to ship without if It's too expensive.

File: extension/app/src/lib/storage.ts [hunk 1]
```diff
@@ -208,3 +208,28 @@ export const getFileContentFragmentId = async (
   const result = await chrome.storage.local.get([key]);
   return result[key] ?? null;
 };
+
+export type SavedAssistantConfiguration = {
```
Comment (PopDaph) [inline extension/app/src/lib/storage.ts:212]: NIT: `AssistantConfiguration` could easily be mislead with `AgentConfiguration` so I would be in favor of renaming for clarity. What about `QuickAction` or event `ContextMenuActions` to be super explicit?

File: extension/app/src/components/conversation/UserMessage.tsx [hunk 2]
```diff
@@ -50,6 +77,86 @@ export function UserMessage({
...
+    (c) => c.id === savedConfigurationId
+  );
+
+  const buttons = isFirstMessage
+    ? [
+        <Button
+          variant="outline"
+          size="xs"
+          icon={enabled ? HeartIcon : HeartStrokeIcon}
+          label={enabled ? "Remove from context menu" : "Save to context menu"}
```
Comment (PopDaph) [inline extension/app/src/components/conversation/UserMessage.tsx:113]: I don't think "context menu" is easily understandable, I suggest "Save as Quick Action" / "Remove from Quick Actions" 🙏🏻

**Review by PopDaph — APPROVED** [review]
Awesome idea 🙌🏻

File: extension/app/src/components/conversation/UserMessage.tsx [hunk 3]
```diff
@@ -50,6 +77,86 @@ export function UserMessage({
...
+              void sendMessage({
+                type: "UPDATE_SAVED_CONFIGURATIONS",
+                configurations,
+              });
+              setSavedConfigurations(configurations);
+            } else {
+              if (savedConfigurationId) {
+                const text = elRef.current?.innerText || "";
+                const description =
+                  text.length > 30 ? text.substring(0, 30) + "..." : text;
```
Comment (PopDaph) [inline extension/app/src/components/conversation/UserMessage.tsx:128]: NIT: could we do more than 30 chars? 🙏🏻 
Also, there's a special char for ellipsis, you can get it from https://www.compart.com/en/unicode/U+2026

**Comment by spolu** [comment]
@PopDaph as DRI. Are we sure this kind of features have to be in-scope for the release? Was this asked by users? How was it prioritized vs say, being able to do something with 5 open tabs?

Also how do I remove a saved thing? This will surely be annoying to have an old assisstant in there. Which means we're opening the door for a non trivial product surface within the extension here no?

File: extension/app/src/components/conversation/UserMessage.tsx (hunk 3 above)
Comment (tdraier) [reply extension/app/src/components/conversation/UserMessage.tsx:121]: Actually we don't need to add an ellipsis, chrome handles it by utself (smthg around 70 chars)

**Comment by tdraier** [comment]
> @PopDaph as DRI. Are we sure this kind of features have to be in-scope for the release? Was this asked by users? How was it prioritized vs say, being able to do something with 5 open tabs?
> 
> Also how do I remove a saved thing? This will surely be annoying to have an old assisstant in there. Which means we're opening the door for a non trivial product surface within the extension here no?

It was first requested internally ( @thib-martin ), and then just got a similar request from alan . the concepts of "quick actions" was in the original design ( although not completely framed,  translate and summarize : https://www.figma.com/proto/QOxHwppG64GtPocpDhS6Y7/Product?page-id=6856%3A9423&node-id=7190-11764&viewport=6125%2C1345%2C0.42&t=LmsVWAucMQUHeg01-1&scaling=min-zoom&content-scaling=fixed )

**Comment by tdraier** [comment]
> Roger that. It is a bit of a dangerous game to reactively add features before releasing a first basic version as it will with low signal change the behavior of user and bias the feedback we'll get. In particular I thiiiink it was decided early on that the "quick actions" would be kept for a subsequent iteration no?
> 
> Genuine question: what's the plan to remove them?

You can add/remove with the same button (like a favorite)

**Comment by tdraier** [comment]
> > Roger that. It is a bit of a dangerous game to reactively add features before releasing a first basic version as it will with low signal change the behavior of user and bias the feedback we'll get. In particular I thiiiink it was decided early on that the "quick actions" would be kept for a subsequent iteration no?
> > Genuine question: what's the plan to remove them?
> 
> You can add/remove with the same button (like a favorite)

And then as proposed by https://github.com/dust-tt/dust/pull/8880#issuecomment-2500401049 , add settings to manage them

**Comment by tdraier** [comment]
> > You can add/remove with the same button (like a favorite)
> 
> You mean you have to type the same thing and then click?

You can go to any conversation created with the quick action (or the original one) and click on the button

**Comment by spolu** [comment]
> You can go to any conversation created with the quick action (or the original one) and click on the button

I fear this will not pass the bar in terms of product standards but I might be wrong. Was this discussed with @Duncid / @pinotalexandre / @gabhubert ?

**Comment by spolu** [comment]
(Also I'd much rather hear 100s feedbacks saying "I can't do x but I really really want to" so that we build something that we know will solve the issue rather than adding a capability before we have usage)

As an example we never built the equivalent in the main app or the need could be considered existing equally no?

**Comment by PopDaph** [comment]
> @PopDaph as DRI. Are we sure this kind of features have to be in-scope for the release? Was this asked by users? How was it prioritized vs say, being able to do something with 5 open tabs?

Yes it was asked by users, and yes this does not have to be in scope for the release. 
(As of now no one asked to do something with 5 tabs, and this is a use case we technically already support if you click on the extract content button from each of your 5 tabs.)

It was not prioritized there was no card for it, my understanding is it was seen as a quick win from @tdraier to solve Thibaut and Alan's use cases. I think this is the kind of feature that can do the wow effect so I think even though I agree this could have been framed first and it should not block the release, I think it's a great addition to boost usage. 

> Also how do I remove a saved thing? This will surely be annoying to have an old assistant in there. Which means we're opening the door for a non trivial product surface within the extension here no?

I shared the same feedback here https://github.com/dust-tt/dust/pull/8880#issuecomment-2500401049
We will deprio - move back to framing - and implement as an improvement post MVP.

**Comment by spolu** [comment]
> not quite sure it make sense in the main app - as here we're including the content of the page (or selection), this is completely tied to the extension itself

Same need exist with attachments :)

**Comment by spolu** [comment]
Actually meditating on this even more I am even more bearish on this change. This basically introduce a new paradigm that overlaps with assistants. To perform repeatable actions you now have this and creating an assistant.

I would much rather invest time in streamlining the creating of assistants for such actions than introducing a new paradigm that overlaps with the very concept of assistants.

**Comment by Duncid** [comment]
Wow, this feels super advanced. I like the quick actions idea and I find the premise interesting, but why do we want to introduce this now?
- It's very "technical"
- Lot of real estate for a context menu (that probably not many people use)
- Once you add, you need a way to manage and then we'll be nuilding product surface on it

I'd punt this to future releases.

**Pull request closed without merging**